_posts/2022-07-18-azure-ad-authentication-cypress.md

```diff
@@ -0,0 +1,185 @@
+
+- `idTokenClaims` - this is the payload of the JWT token which contains information about the user logging in and additional data
+- `localAccountId` - is used to identify a user within the the Azure tenant
+- `homeAccountId` - is used to uniquely identify a user in across all of Azure Active Directory
+
+```javascript
+const injectTokens = (tokenResponse: ExternalTokenResponse) => {
+ const environment = 'login.windows.net';
+ const idTokenClaims: JwtPayload = decode(tokenResponse.id_token);
+ const localAccountId = idTokenClaims.oid || idTokenClaims.sid;
+ const clientId = Cypress.env('AZURE_CLIENT_ID');
+ const realm = Cypress.env('AZURE_TENANT_ID');
+ const homeAccountId = `${localAccountId}.${realm}`;
+ ...
+}
+```
+
+Using these values, we can reconstruct the session storage items that `@azure/msal-browser` expects to see when checking if a user is already logged in. We begin by assembling the session item key as follows.
+
+```javascript
+const tokenId = `${homeAccountId}-${environment}-${realm}`;
+```
+
+The session item value consists of the deconstructed token, some information about the Azure AD tenant and also the user. We then store that value in session storage.
+
+```javascript
+const token = {
+ authorityType: 'MSSTS',
+ homeAccountId,
+ environment,
+ realm,
+ idTokenClaims,
+ localAccountId,
+ username: idTokenClaims.preferred_username,
+ name: idTokenClaims.name,
+};
+sessionStorage.setItem(tokenId, JSON.stringify(token));
+```
+
+We follow a similar pattern for the `accesstoken` session item which includes details about the token expiry, session scopes and the token itself.
+
+```javascript
+const accessTokenId = `${homeAccountId}-${environment}-accesstoken-${Cypress.env(
+ 'AZURE_CLIENT_ID'
+)}-${Cypress.env('AZURE_TENANT_ID')}-${tokenResponse.scope}--`;
+
+const now = Math.floor(Date.now() / 1000);
+const accessToken = {
+ credentialType: 'AccessToken',
+ tokenType: 'Bearer',
+ homeAccountId,
+ secret: tokenResponse.access_token,
+ cachedAt: now.toString(),
+ expiresOn: (now + tokenResponse.expires_in).toString(),
+ extendedExpiresOn: (now + tokenResponse.ext_expires_in).toString(),
+ environment,
+ target: tokenResponse.scope,
+ realm,
+ clientId,
+};
+sessionStorage.setItem(accessTokenId, JSON.stringify(accessToken));
+```
+
+Note that in the screenshot above, we had 4 items in session storage but we have only defined 2 in our example; this is because the additional items aren't needed for the duration of our tests. One of the missing items is the refresh token that is only needed when the token expires and the other contains much of the same information as our `accesstoken` session item.
+
+## 3. Run the test ✅
+
+We can now write a test that demonstrates the login functionality as shown below.
+
+```javascript
+describe('login spec', () => {
+ before(() => {
+  cy.login();
```

Comment:
You should mention in either a new section or in a paragraph the new `cy.session` implementation which manages your sessions in cache, so you don't need to keep logging in OR manage authentication yourself (it does it for you). It naturally will speed up tests as well.

Reply:
I think I might leave out the session stuff for a possible follow up, in their documentation it says its still in an experimental phase and I don't know enough about it to comment at this point.